include/uk/arch/spinlock.h

```diff
@@ -21,20 +23,62 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
-/* Taken from Mini-OS */
 
 #ifndef __UKARCH_SPINLOCK_H__
 #define __UKARCH_SPINLOCK_H__
 
+#include <uk/config.h>
+#include <stdint.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef struct {} spinlock_t;
+#ifdef CONFIG_UKPLAT_LCPU_MULTICORE
+/** spinlock struct */
+typedef struct {
+	volatile uint32_t lock;
+} __spinlock_t;
+
+/** section attribute cannot be used with pointers */
+typedef __spinlock_t *spinlock_ptr_t;
+
+/* spinlock type for declaration */
+#define spinlock_t __attribute__((section(".slocks"))) __spinlock_t
+
+/**
+ * Initialize a spinlock object (put in unlocked state)
+ * @param [in/out] lock Pointer to spinlock object
+ */
+void ukarch_spin_lock_init(spinlock_ptr_t spinlock);
+
+/**
+ * Acquire lock. It is guaranteed that the spinlock object will be locked
+ * in a atomic way, so no other function can have the same object in a unlocked
+ * state.
+ * @param [in/out] lock Pointer to spinlock object
+ */
+void ukarch_spin_lock(spinlock_ptr_t lock);
+
+/**
+ * Release lock previously acquired by spin_lock.
+ * Use store-release to unconditionally clear the spinlock variable.
+ * Store operation generates an event to all cores waiting in WFE
+ * when address is monitored by the global monitor.
+ * @param [in/out] lock Pointer to spinlock object
+ */
+void ukarch_spin_unlock(spinlock_ptr_t lock);
+
+/* Defines a pre-initialized spin_lock in unlocked state */
```

Comment:
```suggestion
/* Defines a pre-initialized spinlock in unlocked state */
```

Also fix in the non-SMP branch.